JM/CF/Scripts/3_Game/CommunityFramework/Timer/CF_TimerBase.c

```diff
@@ -0,0 +1,202 @@
+class CF_TimerBase : Managed
+{
+	private static ref CF_TimerBase s_Root;
+	private ref CF_TimerBase m_Next;
+	private CF_TimerBase m_Prev;
+	private bool m_IsActive;
+	private bool m_Destroy;
+
+	protected float m_Interval; // Seconds
+	protected float m_DeltaTime; // Seconds
+	protected int m_Time; // Milliseconds
+
+	protected void CF_TimerBase()
+	{
+		#ifdef CF_TRACE_ENABLED
+		CF_Trace trace(this, "CF_TimerBase");
+		#endif
+
+		OnCreate();
+	}
+
+	void ~CF_TimerBase()
+	{
+		#ifdef CF_TRACE_ENABLED
+		CF_Trace trace(this, "~CF_TimerBase");
+		#endif
+
+		Stop();
+
+		OnDestroy();
+	}
+
+	void Start()
+	{
+		#ifdef CF_TRACE_ENABLED
+		CF_Trace trace(this, "Start");
+		#endif
+
+		if (m_IsActive) return;
+
+		m_IsActive = true;
+
+		if (s_Root == null)
+		{
+			s_Root = this;
+		}
+		else
+		{
+			m_Next = s_Root;
+			s_Root.m_Prev = this;
+			s_Root = this;
+		}
+
+		OnStart();
+	}
+
+	/**
+	 * Stops the timer, decrementing the reference count.
+	 */
+	void Stop()
+	{
+		#ifdef CF_TRACE_ENABLED
+		CF_Trace trace(this, "Stop");
+		#endif
+
+		if (!m_IsActive) return;
+
+		m_IsActive = false;
+
+		if (s_Root == this)
+		{
+			s_Root = m_Next;
+			if (s_Root) s_Root.m_Prev = null;
+		}
+		else
+		{
+			m_Prev.m_Next = m_Next;
+			if (m_Next) m_Next.m_Prev = m_Prev;
+		}
+
+		OnStop();
+	}
+
+	void Destroy()
+	{
+		#ifdef CF_TRACE_ENABLED
+		CF_Trace trace(this, "Stop");
+		#endif
+
+		m_Destroy = true;
+	}
+
+	void SetInterval(float duration)
+	{
+		#ifdef CF_TRACE_ENABLED
+		CF_Trace trace(this, "SetInterval", "" + duration);
+		#endif
+
+		m_Interval = duration;
+		if (m_Interval < 0) m_Interval = 0;
+	}
+
+	float GetInterval()
+	{
+		#ifdef CF_TRACE_ENABLED
+		CF_Trace trace(this, "GetInterval");
+		#endif
+
+		return m_Interval;
+	}
+
+	void SetTime(int time)
+	{
+		#ifdef CF_TRACE_ENABLED
+		CF_Trace trace(this, "SetTime", "" + time);
+		#endif
+
+		m_Time = time;
+	}
```

Comment:
Setting the elapsed time ... why? Resetting a timer sure, but then I would only add a Reset() function tbh ... setter feels odd here. It implies that setting values other than 0 are intentional by design, which I have not been able to come up with a use-case for.

Reply:
> which I have not been able to come up with a use-case for.

Syncing the world event time from the server to newly connected clients

Reply:
Though this sync needs some extra care by transmitting simulation steps. If you just send the ms count via RPC it will be out of sync by the time it arrives. *Maybe* this problem is worth providing a built-in solution as a wrapper on top or static functionality of a `CF_NetTimer` `CF_SyncedTimer` ... something like that.